### PR TITLE
[REVIEW] Fix `codecov` in CI

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR fixes `codecov` reports generation when used with `pytest-xdist`.

### CODE REVIEW

File: ci/gpu/build.sh [hunk 1]
```diff
@@ -239,9 +239,9 @@ fi
 # TEST - Run py.test, notebooks
 ################################################################################
 
-cd "$WORKSPACE/python/cudf"
+cd "$WORKSPACE/python/cudf/cudf"
 gpuci_logger "Python py.test for cuDF"
-py.test -n 8 --cache-clear --basetemp="$WORKSPACE/cudf-cuda-tmp" --ignore="$WORKSPACE/python/cudf/cudf/benchmarks" --junitxml="$WORKSPACE/junit-cudf.xml" -v --cov-config=.coveragerc --cov=cudf --cov-report=xml:"$WORKSPACE/python/cudf/cudf-coverage.xml" --cov-report term --dist=loadscope cudf
+py.test -n 8 --cache-clear --basetemp="$WORKSPACE/cudf-cuda-tmp" --ignore="$WORKSPACE/python/cudf/cudf/benchmarks" --junitxml="$WORKSPACE/junit-cudf.xml" -v --cov-config=../.coveragerc --cov=cudf --cov-report=xml:"$WORKSPACE/python/cudf/cudf-coverage.xml" --cov-report term --dist=loadscope cudf
```
Comment (bdice) [inline ci/gpu/build.sh:244]: Should `--cov-config` use an absolute path like `"$WORKSPACE/python/cudf/.coveragerc"`, to match the style of the other paths in this command?

File: python/cudf/.coveragerc [hunk 2]
```diff
@@ -1,4 +1,4 @@
 # Configuration file for Python coverage tests
 [run]
-include = cudf/*
-omit = cudf/tests/*
\ No newline at end of file
+source = cudf/*
```
Comment (bdice) [inline python/cudf/.coveragerc:3]: Alternatively, you can just specify the package name:
```suggestion
source = cudf
```
> You can specify source to measure with the `--source` command-line switch, or the `[run] source` configuration value. The value is a comma- or newline-separated list of directories or importable names (packages or modules).

https://coverage.readthedocs.io/en/6.3.2/source.html#execution

File: ci/gpu/build.sh [hunk 3]
```diff
@@ -239,9 +239,10 @@ fi
 # TEST - Run py.test, notebooks
 ################################################################################
 
-cd "$WORKSPACE/python/cudf"
+cd "$WORKSPACE/python/cudf/cudf"
+# It is essential to cd into $WORKSPACE/python/cudf/cudf as `pytest-xdist` + `codecov` seem to work only at this directory level.
```
Comment (vyasr) [inline ci/gpu/build.sh:243]: Yes, coverage.py is extremely sensitive to file paths unfortunately.

**Review by vyasr — APPROVED** [review]
This looks great to me! It may also be nice at some point to enable the `cython.Coverage` plugin so that we can get coverage results for our Cython code, but the pending switch to scikit-build will bork that anyway so let's not bother with enabling coverage until we can improve scikit-build/coverage compatibility.

File: ci/gpu/build.sh (hunk 3 above)
Comment (bdice) [inline ci/gpu/build.sh:243]: `coverage` is the plugin that measures coverage. `codecov` is just the service we use to upload and visualize the results.
```suggestion
# It is essential to cd into $WORKSPACE/python/cudf/cudf as `pytest-xdist` + `coverage` seem to work only at this directory level.
```

File: python/cudf/.coveragerc [hunk 4]
```diff
@@ -1,4 +1,4 @@
 # Configuration file for Python coverage tests
 [run]
-include = cudf/*
-omit = cudf/tests/*
\ No newline at end of file
+source = cudf
+omit = cudf/tests/*
```
Comment (bdice) [inline python/cudf/.coveragerc:4]: Do we really want to omit our tests? Unused functions in our tests should be flagged as "not covered."

**Review by bdice — COMMENTED** [review]
Are any changes needed to measure coverage for dask-cudf and custreamz, or are those already working? Let's try to align coverage for all the Python package tests in this PR if possible.

**Review by Ethyling — APPROVED** [review]

**Comment by galipremsagar** [comment]
> Are any changes needed to measure coverage for dask-cudf and custreamz, or are those already working?

Yea, they seem to be already working.

**Comment by bdice** [comment]
For this PR, I see https://github.com/rapidsai/cudf/blob/branch-22.04/python/dask_cudf/.coveragerc is skipping coverage for dask_cudf tests. I would recommend the same change we made for cudf, don't ignore coverage in test files.

For a follow-up, it looks like `cudf/_fuzz_testing` is not covered (not executed?) in our CI. Should we be running the fuzz testing code (very briefly?) to ensure it works and gives us a coverage measurement during our CI runs?

**Review by bdice — APPROVED** [review]

File: python/dask_cudf/.coveragerc [hunk 5]
```diff
@@ -1,4 +1,3 @@
 # Configuration file for Python coverage tests
 [run]
 include = dask_cudf/*
```
Comment (bdice) [inline python/dask_cudf/.coveragerc:3]:
```suggestion
source = dask_cudf
```

Comment (bdice) [reply python/dask_cudf/.coveragerc]: To align with previous conversation: https://github.com/rapidsai/cudf/pull/10347#discussion_r813554501

**Comment by galipremsagar** [comment]
> For this PR, I see https://github.com/rapidsai/cudf/blob/branch-22.04/python/dask_cudf/.coveragerc is skipping coverage for dask_cudf tests. I would recommend the same change we made for cudf, don't ignore coverage in test files.
> 

Made the changes.

> For a follow-up, it looks like `cudf/_fuzz_testing` is not covered (not executed?) in our CI. Should we be running the fuzz testing code (very briefly?) to ensure it works and gives us a coverage measurement during our CI runs? I am not familiar with this part of cudf so feel free to ignore if I'm missing something here (e.g. if the code is not stable or can't be executed on CI).

Those are very long-running tests as they target at the machine limits dataset sizes. Hence they are not intended to be run per PR. So we are planning on running them on benchmark machines soon.

**Comment by bdice** [comment]
> Those are very long-running tests as they target at the machine limits dataset sizes. Hence they are not intended to be run per PR. So we are planning on running them on benchmark machines soon.

Right, I know fuzz testing is typically large-scale. I was wondering if there are options to use "toy" dataset sizes that can run quickly, to permit CI testing that ensures the code _works_ and has reasonable coverage.

**Comment by galipremsagar** [comment]
> > Those are very long-running tests as they target at the machine limits dataset sizes. Hence they are not intended to be run per PR. So we are planning on running them on benchmark machines soon.
> 
> Right, I know fuzz testing is typically large-scale. I was wondering if there are options to use "toy" dataset sizes that can run quickly, to permit CI testing that ensures the code _works_ and has reasonable coverage.

Yeah, I think that is reasonable and possible, I'll try this in a follow-up PR.

**Review by quasiben — APPROVED** [review]

**Comment by quasiben** [comment]
@gpucibot merge

**Pull request merged**